keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1561 @@
+
+```yaml
+annotations:
+  evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS_HASH}: ${PRIORITY}/${ROLE}
+  evacuation.coordination.k8s.io/hash_${EVACUATOR_CLASS_HASH}: ${EVACUATOR_CLASS}
+```
+
+which would result in
+
+```yaml
+annotations:
+  evacuation.coordination.k8s.io/priority_7bb77f87dc: "11000/knowledgeable-app-specific"
+  evacuation.coordination.k8s.io/hash_7bb77f87dc: "sensitive-workload-operator.tom-and-jerry-delicious-fruit-company.com"
+```
+
+`7bb77f87dc` value is a hash of `sensitive-workload-operator.tom-and-jerry-delicious-fruit-company.com`
+
+### Risks and Mitigations
+
+If there is no evacuator, and the application has insufficient availability and a blocking PDB or
+blocking validating admission webhook, then the evacuation controller will enter into an eviction
+cold loop. To mitigate this we will increment `.status.failedEvictionCounter` and
+`evacuation_controller_evictions` metric.
+
+An evacuator could reconcile the status properly without making any progress. It is thus
+recommended to check `creationTimestamp` of the Evacuations and observe
+`evacuation_controller_active_evacuator` metric to see how much it takes for an evacuator to
+complete the evacuation. This metric can be also used to implement additional alerts.
+
+#### Disruptive Eviction
+
+When using kubectl drain, pods without owning controller and pods with local storage
+(having `emptyDir` volumes) are not evicted by default. We have decided to evict most
+of the pods (except DaemonSet and mirror pods) by default. In the motivation section of the
+[Declarative Node Maintenance KEP](https://github.com/kubernetes/enhancements/pull/4213),
+we can see many administrators override these default settings and many components evict all pods
+indiscriminately. There are also many ways that users use to prevent the eviction;
+PodDisruptionBudgets, validating admission webhooks, or just plain HA. Users who want to protect
+their applications in today's clusters should already be aware that they should be able to handle
+sudden evictions. Therefore, it should be okay for the evacuation controller to evict these pods.
+
+To mitigate the sudden eviction problem, users should use PodDisruptionBudgets or HA.
+
+## Design Details
+
+### Evacuation
+
+We will introduce a new type called `evacuation.coordination.k8s.io`  to enforce the contract
+between the evacuation instigators and the evacuators. This type is a bit similar to
+`leases.coordination.k8s.io` in that it requires multiple actors to synchronize the state. Which in
+our case is the progress of the evacuation.
+
+### Evacuation Instigator
+[Evacuation Instigator](#evacuation-instigator) section provides a general overview.
+
+There can be many evacuation instigators for a single Evacuation.
+
+When an instigator decides that a pod needs to be evacuated, it should create an Evacuation:
+- With a predictable and consistent name in the following format: `${POD_UID}-${POD_NAME_PREFIX}`.
```

Comment:
I don't love magical hardcoded formatted strings.  Why is it not sufficient to use POD_NAME or, if you really want to handle the ABA case, just POD_UID ?  What I mean is that we already have places that assert that two things are linked 1:1 by having the same name.  In this case, I suppose it's possilbe that you have:

- Pod "foo" exists
- create evacuation "foo", with pod's UID inside
- delete pod foo
- create pod foo
- try to create evacuation "foo", with new UID inside, fail

I (maybe wrongly) presume there's a cleanup loop that removes evacuations which have no matching pod?

Reply:
I have analysed this a bit more and decided to use solely POD_UID as the allowed name, you can see the reasoning behind this in https://github.com/atiratree/kube-enhancements/blob/evacuation-api/keps/sig-apps/4563-eviction-request-api/README.md#the-name-of-the-evictionrequest-objects